File: LeetCode_Java/stack/MinimumRemoveToMakeValidParentheses.java
package stack;

import java.util.HashSet;
import java.util.Set;
import java.util.Stack;

public class MinimumRemoveToMakeValidParentheses {
    public String minRemoveToMakeValid(String s) {
        Stack<Integer> match = new Stack<>();
        Set<Integer> notMatch = new HashSet<>();

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '(') {
                match.push(i);
            } else if (ch == ')') {
                if (!match.isEmpty()) {
                    match.pop();
                } else {
                    notMatch.add(i);
                }
            }
        }

        while (!match.isEmpty()) {
            notMatch.add(match.pop());
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (!notMatch.contains(i)) {
                char ch = s.charAt(i);
                sb.append(ch);
            }
        }

        return sb.toString();
    }
    public static void main(String[] args) {
        MinimumRemoveToMakeValidParentheses solution = new MinimumRemoveToMakeValidParentheses();
        System.out.println(solution.minRemoveToMakeValid("lee(t(c)o)de)"));// Output: "lee(t(c)o)de"
        System.out.println(solution.minRemoveToMakeValid("a)b(c)d"));// Output: "ab(c)d"
        System.out.println(solution.minRemoveToMakeValid("))(("));// Output: ""
    }
}

File: LeetCode_Python/stack/MinimumRemoveToMakeValidParentheses.py
class MinimumRemoveToMakeValidParentheses:
    def minRemoveToMakeValid(self, s: str) -> str:
        arr = list(s)
        stack = []

        for i, char in enumerate(arr):
            if char == "(":
                stack.append(i)
            elif char == ")":
                if stack:
                    stack.pop()
                else:
                    arr[i] = ""

        while stack:
            arr[stack.pop()] = ""

        return "".join(arr)


# Time Complexity: O(n)
# Space Complexity: O(n)

if __name__ == "__main__":
    solution = MinimumRemoveToMakeValidParentheses()

    s = "lee(t(c)o)de)"
    print("Expected Output : lee(t(c)o)de")
    print("Actual Output :", solution.minRemoveToMakeValid(s))

    s = "a)b(c)d"
    print("Expected Output : ab(c)d")
    print("Actual Output :", solution.minRemoveToMakeValid(s))

    s = "))(("
    print("Expected Output : ")
    print("Actual Output :", solution.minRemoveToMakeValid(s))

    s = "(a(b(c)d)"
    print("Expected Output : a(b(c)d)")
    print("Actual Output :", solution.minRemoveToMakeValid(s))
